### ISSUE TITLE
The ongoing, everlasting, LibreQos design meeting

### ISSUE THREAD

**Issue by dtaht** [LibreQoE/LibreQoS#57]
@interduo @richb-hanover @Belyivulk @tohojo @rchac @mjstekel @syadnom @mwangi-kiboi @axonxorz @thebracket @marsalans @micron10  (and whoever else I missed) - I have been watching this project expand in various directions rapidly and I hope, meeting many of your pent-up needs in production. I have some spare time in the coming quarter, and trying to get a grip on what is important, what is needed, what resources are missing, getting some bugs closed, and having an overall architectural discussion, and a sane roadmap is on my mind. 

A) I am in PST, and available 7-7 for a videoconference any time next week wed-friday. Needn't be with all of you at the same time... 
B) is anyone here going to wispaloosa? 

I've mostly been observing a big wireless-n network and looking over the cake statistics deeply, of late. I have a few thoughts towards improving cake, I've seen some thoughts towards improving the htb stuff, ipv6 support is still a pita, and I don't know what else.

**Comment by rchac** [comment]
@dtaht Great idea. I think it would be awesome to work out the highest priority features and bug fixes, and develop a roadmap. 8:00 AM PST on Wednesday, September 7th would be pretty ideal for me. Would anyone else be interested?

**Comment by Belyivulk** [comment]
That is 3am for me here 😲 I am far too old for 3am’s!

HAYDEN​​ SIMON
UBER GROUP LIMITED
MANAGING DIRECTOR
 ‑
SUPREME OVERLORD
[http://uber.co.nz/wp-content/uploads/signatures/uber-facebook-icon.png]<https://www.facebook.com/UberGroup?_rdr=p>
[http://uber.co.nz/wp-content/uploads/signatures/uber-twitter-icon.png]<https://twitter.com/ubergroupltd>
E: ***@***.******@***.***>
M: 021 0707 014<tel:021%200707%20014>
W: www.uber.nz<http://www.uber.nz/>
53 PORT ROAD | PO BOX 5083 | WHANGAREI | NEW ZEALAND

**Comment by tohojo** [comment]
I'll be busy preparing for, and then going to, LPC, so y'all go ahead and meet without me. Happy to join sometime after that, though!

**Comment by rchac** [comment]
@Belyivulk 

Sorry about that. More times next week:
- Tuesday the 6th - 8AM PST to 12PM PST
- Tuesday the 6th - 3PM PST to 7PM PST
- Thursday the 8th - 4PM PST to 7PM PST
- Friday the 9th - 4PM PST to 7PM PST

Would any of these be better?

**Comment by thebracket** [comment]
I'm traveling to visit family overseas for all of those dates. I'll try and
get you access to a sanitized bracketqos; it's way too coupled with our
company infrastructure to be shared as-is (but is working wonderfully for a
52 tower wisp!)

**Comment by dtaht** [comment]
@thebracket - how do you know? :) I'm a statistics junky, and care about observing drops/marks/backlog and some of the other statistics in cake like pk_delay, especially at peak hours... I'm seeing a truly astonishing amount of marks, are you tracking that?

Anecdotes cheer me up, too. One anecdote I got back was: "We deployed libreqos, and overnight all our customer support complaints about speed *vanished*, about 8% of call volume. We then set to work reflashing a bunch of old cpe to openwrt to distribute to other customers."

**Comment by dtaht** [comment]
In general I have found that around 1PST suitable for concalls between NZ and Denmark. I would be avail for your 8AM wednesday PST robert, as it's long past time we had some short RTTs between us. Perhaps if we shoot for a meeting week after this?

Also, would people find a mailing list to be of benefit?

**Comment by rchac** [comment]
@dtaht That'd be great. Let's have a call at that time on Wednesday this week.

Maybe a more general meeting the following week would work better for everyone? For example on the 14th and 15th I'm clear from 12PM PST until the end of the day. As Dave suggested we could try for around 1PM PST?

**Comment by thebracket** [comment]
> @thebracket - how do you know? :) I'm a statistics junky, and care about observing drops/marks/backlog and some of the other statistics in cake like pk_delay, especially at peak hours... I'm seeing a truly astonishing amount of marks, are you tracking that?
> 
> Anecdotes cheer me up, too. One anecdote I got back was: "We deployed libreqos, and overnight all our customer support complaints about speed _vanished_, about 8% of call volume. We then set to work reflashing a bunch of old cpe to openwrt to distribute to other customers."

Always a good question. The anecdote version:

* We had a Preseem instance, serving around 1-1.25 gbps at peak time (with a low closer to 400mbps). Prior to Preseem, we had a manageable but small number of slow speed calls - but VOIP users and gamers complained quite a bit. Once we installed Preseem, these complaints dropped significantly.
* We had some issues with a Preseem upgrade that prompted us to look elsewhere. Not going to point fingers.
* Since I was on a 12-week paternity leave with a baby at home, I rewrote LibreQOS entirely in Rust, added a full web stack, integrated it with our UISP, cnMaestro and Radius setup for some wifi and Mimosa clients (we've acquired 2 WISPs over the years and have a wide range of gear). It retains the multi-CPU queues, the basic queue structure and TCP latency tracking - in other words, the core is all there, it's just in a language I love using.
* Annoyed at Preseem, we decided to cut-over to my version - *BracketQOS* - for a test. It worked *really* well, and the VOIP users and gamers were now happy! So we ditched Preseem.

There's about 600 devices being tracked right now. It still gets up to about 1.25 Gbps at peak, but our average has actually lowered to closer to 900 for peak nights; despite the reduced bandwidth usage, we've actually had very, very few speed complaints (fewer than with Preseem) - and they almost always turn out to be "we weren't tracking that device properly".

We arrange our queues with a top-level interface queue, a queue per site, queues for APs at sites, and "backhaul queues" limiting downstream towers that are linked together - fairly sharing limited backhaul.

The front page (3 month view, which is averaging stats downwards a bit) looks like this:
![image](https://user-images.githubusercontent.com/14896751/188342081-c40be41f-c4db-4d0c-9894-d8cd071bfa1f.png)

The latency graph shows median as the line and best in the light blue at the bottom, worst (whole network) on the top. So we can tell at a glance that TCP latency is comfortable in aggregate (it's floating around 70ms network-wide right now). The bottom graphs show stats for site queues, giving an at-a-glance "where is my bandwidth going?" chart.

So zooming into a busy site (with some names blurred, because there's customer names visible), and looking at it right now:
![image](https://user-images.githubusercontent.com/14896751/188342310-5574813d-e744-49a7-8633-fa9d3d7ba8ab.png)

The site has some backhaul issues (waiting on a climber), so the site has a limit of 400mbps set. That's great, because the queues are compensating. Latency remains consistently good (50-60ms average, 150ms worst-case), throughput is constrained (despite this site feeding 4 micro-pops as well as a bunch of its own APs). Preseem wouldn't even let me do that.

So zooming down to a customer who used to complain a lot (more pixelization for names):
![image](https://user-images.githubusercontent.com/14896751/188342624-2c4fa208-190c-4cd9-9af5-88dfb8bd352f.png)

This customer pays for a 20/5 - for historical reasons, they are grandfathered in and don't want to go with a bigger plan despite filling their pipe every single night. It's a PMP450b, hanging off of a Medusa on the CALVIN site. As you can see, they will consume their full allocation for hours at a time. You can see "drops" graphed (we don't track maps), latency remains fantastic - even thought the tower is approaching backhaul capacity, and they are using their entire connection (and a tiny bit more, occasionally - which we're cool with).

And here's the raw qdisc stats for that customer's queue:

```
qdisc cake d410: parent 1:41 bandwidth unlimited diffserv4 triple-isolate nonat nowash no-ack-filter split-gso rtt 100ms raw overhead 0
 Sent 710197820346 bytes 547146733 pkt (dropped 15543770, overlimits 0 requeues 0)
 backlog 0b 0p requeues 0
 memory used: 3896064b of 15140Kb
 capacity estimate: 0bit
 min/max network layer size:           56 /    1514
 min/max overhead-adjusted size:       56 /    1514
 average network hdr offset:           14

                   Bulk  Best Effort        Video        Voice
  thresh           0bit         0bit         0bit         0bit
  target            5ms          5ms          5ms          5ms
  interval        100ms        100ms        100ms        100ms
  pk_delay       14.4ms       2.07ms       3.97ms       16.9ms
  av_delay       5.57ms        584us        662us        6.5ms
  sp_delay         11us          2us          2us          1us
  backlog            0b           0b           0b           0b
  pkts           935274    554142759      4086596      3525874
  bytes      1405288236 724102950561   5549927455   1707069097
  way_inds          779     41473923         1751          484
  way_miss          231      1813231        16007         4555
  way_cols            0            0            0            0
  drops           97352     15259925       181275         5218
  marks               0            0            0            0
  ack_drop            0            0            0            0
  sp_flows            0            2            0            1
  bk_flows            0            1            0            0
  un_flows            0            0            0            0
  max_len          1514         1514         1514         1514
  quantum          1514         1514         1514         1514
```

And in the other direction:

```
qdisc cake d413: parent 1:41 bandwidth unlimited diffserv4 triple-isolate nonat nowash no-ack-filter split-gso rtt 100ms raw overhead 0
 Sent 76629811232 bytes 329167524 pkt (dropped 49906, overlimits 0 requeues 0)
 backlog 0b 0p requeues 0
 memory used: 707328b of 15140Kb
 capacity estimate: 0bit
 min/max network layer size:           60 /    1514
 min/max overhead-adjusted size:       60 /    1514
 average network hdr offset:           14

                   Bulk  Best Effort        Video        Voice
  thresh           0bit         0bit         0bit         0bit
  target            5ms          5ms          5ms          5ms
  interval        100ms        100ms        100ms        100ms
  pk_delay          0us       7.11ms       1.15ms       3.18ms
  av_delay          0us        558us        175us        595us
  sp_delay          0us          3us          2us          1us
  backlog            0b           0b           0b           0b
  pkts                0    213598032        91684    115527714
  bytes               0  66107672378      6275849  10579514811
  way_inds            0     10302772          220     25630838
  way_miss            0      1341744        40004       308233
  way_cols            0            0            0            0
  drops               0        47661            3         2242
  marks               0            0            0            0
  ack_drop            0            0            0            0
  sp_flows            0            2            1            6
  bk_flows            0            0            0            0
  un_flows            0            0            0            0
  max_len             0         1514          170         1514
  quantum          1514         1514         1514         1514
```

Lastly, the TCP latency tracking does eat a bit of CPU - but the system as a whole averages <30% CPU time per CPU for the ones that aren't running `pping`.

Hope that helps!

**Comment by thebracket** [comment]
I went digging for marks (I'm honestly not quite sure what they are telling me). Found a customer with quite a few:

```
qdisc cake d5a2: parent 6:156 bandwidth unlimited diffserv4 triple-isolate nonat nowash no-ack-filter split-gso rtt 100ms raw overhead 0
 Sent 116826433200 bytes 98300622 pkt (dropped 2114568, overlimits 0 requeues 0)
 backlog 0b 0p requeues 0
 memory used: 11996928b of 15140Kb
 capacity estimate: 0bit
 min/max network layer size:           56 /    1514
 min/max overhead-adjusted size:       56 /    1514
 average network hdr offset:           14

                   Bulk  Best Effort        Video        Voice
  thresh           0bit         0bit         0bit         0bit
  target            5ms          5ms          5ms          5ms
  interval        100ms        100ms        100ms        100ms
  pk_delay       13.5ms       26.2ms        4.3ms       2.54ms
  av_delay       1.18ms       15.4ms        263us        622us
  sp_delay         13us          1us          4us          3us
  backlog            0b           0b           0b           0b
  pkts            12836     97034333      3037951       330070
  bytes        13629221 117679145674   1584071518    486263746
  way_inds           13      4577248        77943           48
  way_miss          284       440111        11626          345
  way_cols            0            0            0            0
  drops             108      2087600        20508         6352
  marks               0      1779998          146            0
  ack_drop            0            0            0            0
  sp_flows            1            1            0            0
  bk_flows            0            1            0            0
  un_flows            0            0            0            0
  max_len          1514         1514         1514         1514
  quantum          1514         1514         1514         1514
```

![image](https://user-images.githubusercontent.com/14896751/188345242-d08b8fb8-90c1-4de4-9a3d-7dde9ef4f13a.png)
He's an odd one. Running an ancient Elevated Ubiquiti device in 2.4ghz, on an equally ancient Cambium ePMP 1000. Has a 10/5 plan, and doesn't do very much - one of the "check my email, let the kids play when they visit" type of customers. Lots of drops, latency remains very good. Gets what he's paying for.

**Comment by dtaht** [comment]
Thx for sharing all that. I started an ISP in 1993, sold it in 1999, tried to get a wisp off the ground 2006-2008, ran smack dab into bufferbloat in rain, went under, more details here: https://forum.mikrotik.com/viewtopic.php?t=181289#p899650

Normally I like to compare hour to hour (peak, 8PM) over time and the same day vs the same day, can influxdb do that?

While 60-150ms is remarkably good for a wireless network, it's still more than I was aiming for with codel, but I think factored into that number for you is plus your typical RTT outside your network? Codel aims for 5ms additional on top of your baseline rtt.

As for RFC3168 ECN marks are coming from A) mismarked packets from various sites, mostly in argentina, when last I looked. B) apple devices negotiating it for some protocols and C) people that have explicitly enabled it. lossless congestion control
is theoretically a good thing but deployment numbers have been hard to come by. A simple metric is how many customers have marked traffic in 24 hrs in each direction, then reduced to a percentage to not reveal the number of customers. Another is the drop/mark ratio by customer as a histogram. 

Are you doing dpi and/or actively dscp marking traffic? 

Have you deployed cake or fq_codel running natively (unshaped) on the mikrotik gear?

I'm told the new PMP-4500 will have cake.

Have you ever run a flent rrul test across your backbone to a tower? (long enough to show up in the db, also?)

Not sorry for all the questions!

**Comment by dtaht** [comment]
PS Elsewhere we are fiddling with the ack-drop parameter in both directions.

**Comment by interduo** [comment]
@dtaht im interested, just send me the time and link to zoom/jitsi/other conference tool

**Comment by syadnom** [comment]
bracket, this is exactly what I'm looking for.  I'm running preseem now but
the lack of head-end and site shapers is a real hindrances. I've demoed
paraqum and cambium (Bequant) and really run into the same issues.  I
actually don't care much about DPI and application shaping as I'd rather be
more neutral for customers.

The ability to shape to backhaul capacity for a site is critical especially
as our oversells go higher.  ie, we're selling higher peak capacity plans
while not seeing a dramatic increase in typical usage.  So, if I'm selling
out 200Mbps plans to customers on a 400Mbps backhaul that has a high
average of 250Mbps all is well EXCEPT when a couple high demand downloads
come in and we land in congestion on the backhaul.

**Comment by interduo** [comment]
@syadnom You don't use TDMA like NV2 (Mikrotik) or Airmax (Ubiquiti) for links? (this shares bandwidth between customers fairly)

**Comment by dtaht** [comment]
@thebracket I happen to not like python very much (the gil lock is a showstopper for me, always has been), and although I've programmed in about 50 languages so far, my best are c, shell, and sql - which makes me pretty obsolete - still a good debugger, though, I like to think. I've dabbled in go and rust over the past few years, with the addition of generics I'm finally interested in learning one or the other, it was the quality of the netlink libs and low level abstractions like TCP_INFO and cmsg was rather poor in both languages when last I looked. I REALLY liked tokio. Anyway, another limitation I personally have is I contribute to openwrt and itty bitty devices, still a lot, which keeps my c skills from rotting entirely. 

Anyway, reading this today was very encouraging: https://blog.dend.ro/building-rust-for-routers/

He'd started with a 8MB binary for essentially "hello, world", and got it down to 1.4MB. 

Seeing something like bracket/libreqos work directly on routers would rock my world!

**Comment by syadnom** [comment]
Sure, but it has nothing to do with AQM/Shaping.

**Comment by thebracket** [comment]
If you're willing to go a little crazy, you can get to 608 bytes with a
syscall, strip and marking parts of Rust as unreachable.
https://godbolt.org/z/dxDncd

I tend to write everything in Rust, because writing Rust books (Hands-on
Rust, Rust Brain Teasers, another one in the works) is a big chunk of my
livelihood - and I love tech writing. It's a great language. I haven't done
much tiny embedded yet, but I've made a bunch of small tokio servers that
just sit and work. Once you get into "no standard library" sized embedded,
life gets pretty tricky.

When I return from my travels, I'll try to get a sanitized bracketqos
posted for you guys to take apart. My next goal is to learn BPF (there's
even rust support for it now), and improve the xdp mapping side of things.
I hacked together a version that replaced the flat ip map with cidr support
(which would also be a first step to ipv6), but had problems getting
mapping working between the BPF modules. Hoping I can figure that out when
I return.

**Comment by thebracket** [comment]
I'm back from a vacation (well, kind-of a vacation; more like an urgent trip to see my unwell mother). I've started the process of putting up an open source version of `bracket-qos`. I still have more docs to write and a few hard-coded paths to fix, but hopefully you find this useful:

https://github.com/thebracket/bqos-oss

**Comment by interduo** [comment]
@thebracket why did You wrote new Rust project instead of contributing here? is there any objective that could not be done in python? 

maaaaan that was a big amount of work to rewrite it

**Comment by rchac** [comment]
@interduo See https://github.com/rchac/LibreQoS/issues/57#issuecomment-1236466462 He's a RUST expert. =) And he made some great design improvements from the looks of it. We will look it over thoroughly and bring these cool design features upstream to LibreQoS. Thank you so much @thebracket for taking the time to make an OSS version!

**Comment by thebracket** [comment]
I write Rust for a living (and books about Rust), and hardly ever touch Python. I also wrote it in a hurry, pretty closely coupled to how our network works; so this is the first release that should work for other networks. I actually had the core and manager running in less than a week, and about 3 weeks of tweaking it before we were able to use it in production. So not all that bad in terms of workload!

The main features I wanted:

* Rust is really, really fast. It downloads our topology and builds the queues extremely quickly, minimizing network burps while it reloads.
* I wanted asynchronous support and multi-threaded execution. You *can* do that in Python, but it gets beyond my "use Python scripts to quickly fix things" level of Python knowledge. The stats collection, network update, latency monitoring, etc. are all threaded.

I probably could have made it all work in Python - but instead of a couple of weeks, I'd have spent a month getting up to speed with the Python world, and then been able to start putting the program together. I'm hoping that there's some useful content here for the main LibreQOS project.

**Comment by thebracket** [comment]
I should also mention that I really suck at HTML/CSS/JavaScript... so while the manager pages are pretty, they are pretty wretched code! I hope to improve that.

**Comment by rchac** [comment]
The entire [bqos-oss](https://github.com/thebracket/bqos-oss) project looks fantastic. The code organization, daemon and manager split, UI - it's all awesome. Super happy we have something to guide the direction of LibreQoS now. Really appreciate your work on this.

**Comment by thebracket** [comment]
You're welcome. My time is a bit limited (baby, new book, another big project), but I'm happy to try and help. I have a few improvements in mind:

* Separate the "topology builder" into its own service, to make it easier to support multiple sources of topology data (including user entered, since not everyone uses UISP). Possibly with a REST API of its own for easy integration.
* Multiple-node support, eventually; we're just getting to the point that we need it, so I'll be coding it to work for us first!
* Some kind of authentication for the manager. I had to remove ours, because it's tied into a single sign-on setup we use across a whole bunch of applications - none of which my employer wants me to share!
* I'm starting to teach myself BPF. I've got a really crappy version running that uses `Trie` maps to handle CIDR rather than single IP addresses (so you can attach `192.168.10.0/24` to a single customer). It needs a *lot* of work before I can unleash it. I'm really tempted to rewrite parts of the `xdp-cpumap` project, mostly because I keep getting confused when I work on it!
* The BPF work is a precursor to properly handling IPv6, which is a wish-list item for us (we are testing it, but it's not live yet and won't be until our upstream supports it)

I also need to move our network towards using the OSS version; we have a few features that aren't in there:

* Our version also connects to our cnMaestro and updates displayed data with Maestro information. I managed to make the Maestro code super un-portable and messy, so porting it over is left for the mysterious future. It's serious "face palm" level of code! (Which coincides with our baby deciding that sleep was for the weak...)
* We have a few odd integration features that probably belong in different projects. We use UISP as a "source of truth" and read it for various other things. For example, we set various wifi MDU residents up as "other devices" with their MAC and *intended* IP addresses (so the QoS picks it up and shapes them) - and feeds Option 82 data into a DHCP configuration. That's lovely to have, but with hindsight I'm not at all sure why I put that into the QoS program. :-|

**Comment by interduo** [comment]
Well in my opinion is that there is no need to double work that CRM does.
Shaper should shape customers and nothing more. 

Graphs should be moved to zabbix. Even when I have not so many data influxdb runs slowly.
There are so many cases that zabbix resolves...especialy in big scale.

![image](https://user-images.githubusercontent.com/17087236/190682194-8d80e9f4-583a-46c2-bb81-292de5871757.png)

**Comment by rchac** [comment]
> * I'm starting to teach myself BPF. I've got a really crappy version running that uses `Trie` maps to handle CIDR rather than single IP addresses (so you can attach `192.168.10.0/24` to a single customer). It needs a _lot_ of work before I can unleash it. I'm really tempted to rewrite parts of the `xdp-cpumap` project, mostly because I keep getting confused when I work on it!
> 
> * The BPF work is a precursor to properly handling IPv6, which is a wish-list item for us (we are testing it, but it's not live yet and won't be until our upstream supports it)

<img src="https://c.tenor.com/bQrjUlPjAnsAAAAC/despicable-me-minions.gif"></img>

If we can get proper IPv6 support in xdp-cpumap-tc that'd be incredible. Keep us updated on your progress and how we can help.

**Comment by tohojo** [comment]
Are you aware of Aya? It's a library for writing BPF natively in Rust (both the userspace handling and the actual BPF code itself)...

https://aya-rs.dev/

**Comment by thebracket** [comment]
@tohojo That looks awesome! I'd seen a few Rust BPF libraries, that one looks really nice! I'll see about putting it to use soon. :-)

In the meantime, I dusted off my C skills (used C since the mid 90s, but not for a few years) and resumed hacking on my `xdp-cpumap` fork. The repo is here: https://github.com/thebracket/xdp-cpumap-tc/tree/trie (I'm working in the `trie` branch right now).

Changes so far:

* If `DEBUG` wasn't defined, debug code wasn't entirely removed - so I fixed that.
* `make clean` actually removes all of the binaries, now.
* Some `.gitignore` stuff to help keep it clean.
* `Makefile` changed to optimize for size, since BPF puts you in a *very* constrained environment (256 byte stack, small total limit)
* The BPF modules' IPv4 array is now a `Trie`, and supports storing `address` and `prefixlen`.
* The command line utility `xdp_iphash_to_cpu_cmdline` supports a new argument, `--prefix`. It defaults to 32 (a single address), but you can specify subnets with `src/xdp_iphash_to_cpu_cmdline --add --ip 192.168.20.0 --classid 1:200 --cpu 0 --prefix 24`.
* The command line `list` function has been updated to list the results. No support for "delete" yet.
* Added a `testbed.sh` shell to set things up on a single NIC and let me count packets in the queues to make sure that it actually matched packets.

The C is a mess at this point, needs cleanup. I'll get `delete` support in there (even though I don't use it!), and clean it up when I'm satisfied that it works. Then I need to setup a proper testbed for it and make sure I'm not doing something stupid.

I'm kinda torn right now: I can either match the original style and send it upstream as a PR, or I can turn it into readable C with lots of comments, headers, and spacing that doesn't set off my OCD - and maintain the fork separately. :-|

**Comment by interduo** [comment]
@thebracket could You do possible for xdp-cpu-map that would be possible to upload many filters by one command?

**Comment by thebracket** [comment]
Guessing your thinking of the latency monitor BPF thing? (Sorry, under
sleeping baby - can't remember the name) I was wondering about that, too.

I see two possibilities; there's a mechanism for chaining programs (so one
calls another), but I'm not sure how that works yet. The other option is to
combine into some big program - ideally also including a list of unmapped
ip addresses, since I really don't like how I currently gather them.


(I'm out of time for today, but my next sprint will be firing up a set of
test VMs to properly validate this)

**Comment by tohojo** [comment]
On the TC hook you can load multiple programs by setting different priorities (you just need to take care with the return code that you don't interrupt the pipeline). On the XDP hook you can use libxdp to achieve the same thing...

**Comment by dtaht** [comment]
A really hard problem with wireless is that bandwidth is the sum of tx + rx, unlike a switched network. Also, many ethernet (switched) device drivers do not multiplex well, either, they may be capable of 1gbit in either direction, but incapable of good performance at both at the same time. The rrul and rtt_fair tests will show this. Most of our work to date can only optimize for one direction, there is no shared state between tx and rx that can account for this. We have mostly tried to reduce multiplexing delay as a function of the number of devices present rather than tackle this big problem, and the ATF code we have for wifi does some of it also. 

Somewhat related to this, in emulating wifi, is there is no shared codel state from cake - htb - pop - etc in the path. Running multiple cake qdiscs in series is not as effective as somehow carrying the codel(cobalt) timestamp and state from start to finish.

There was a really good paper published at sigcomm 2022 - https://dl.acm.org/doi/pdf/10.1145/3544216.3544271 which also references the current state of the art in the "Swift" project in some reference or another. 

At one level I'm hesitant to deluge you with this stuff, as I'm all in favor of incremental progress, and me knowing too much is kind of like asking a millipede how it walks in the first place, and causing freezeup. But just to add to those thousand legs, while I found a htb "crate" for rust, I'm unfond of token buckets and vastly prefer the sort of deficit based scheduling cake does as being a closer match to how the underlying hardware works, perhaps with a per-cpu variable rcu- merge strategy - and there was a VERY long series about how ideas like RCU were incompatible with how rust thinks about things here: https://paulmck.livejournal.com/64209.html which I would rather try to discregard in favor of how rust thinks about things. (e.g. don't read this stuff! Carry on!) 

It has been the locking in a qdisc that has been the real pain point for us for many years. Perhaps it's not as much a problem anymore, but I would recommend constant analysis via https://www.brendangregg.com/flamegraphs.html to make sure the thing can scale.

**Comment by syadnom** [comment]
Dave, just to update you..  most non-ubiquiti and some ubiquiti wireless
systems these days actually are fixed values due to GPS sync.  Anyone using
Cambium has been this way a long time.  New epmp 4x00 is fixed frame length
as is Mimosa A6 and essentially everything in 6Ghz outdoor will have to be
to use the AFC.  These are asynchronous ratios though, so that does present
a challenge/need to be able to configure the ratio for shapers.

Also I should note that even on airmax, I test aggregate speeds and shape
each access point based on that aggregate delivery.  ie, 100Mbps aggregate
will get shaped 80x20 (in preseem) so I'm kinda pretending that these are
fixed ratio as well.  If GPS is enabled I do *both* rx/tx + aggregate and
set the shaper to where the AP doesn't fall over.  Cambium radios are much
cleaner in this regard, the fixed ratios can essentially be trusted.

I don't really have a need outside of corner cases to have a dynamic ratio
on APs.

**Comment by dtaht** [comment]
As a moonshot, one of the things we'd hoped to do was to move cake into user space for vpp/dpdk or into a FPGA or p4. The biggest dependency it had was on the really complicated flow-dissector, embedded in the linux kernel. This project (and open source code) has the potential to fix all that:

https://netdevconf.info/0x15/session.html?Replacing-Flow-Dissector-with-PANDA-Parser 

A couple other performance notes - the invsqrt cobalt/codel/etc depends on is a single fpu instruction - which we can't do in kernelspace, so we emulate it via newton's method, and a table lookup. (in ns3 we just used floating point) Cake itself takes 3 hashes, which ideally could just be passed along from instance to instance, and there are faster hash functions in SSE mode. 

After looking at the performance of even ebpf pping, it seems saner to somehow hook into cake near the ack-filter code, and
find a way to share that info between rx and tx instances, so it could be turned on or off on a per-instance basis.

Despite me saying this, it would be a better idea, to just profile where the bottlenecks are, because inevitably they aren't where you think they are.

**Comment by dtaht** [comment]
@syadnom - thx for the update! Everybody please feel free to update me on the present state of the FWA devices "out there" - I exited this market in 2010, and while I'm sure that FQ of some form made it into many devices, I'm still pretty sure most missed the need for AQM, and less sure that the FQ method for stations and flows was adopted outside of fq_codel and cake: 

http://www.diva-portal.org/smash/get/diva2:1251687/FULLTEXT01.pdf

I'm still in a "trust but verify" mode, and hoping more folk test using the rtt_fair test to 3 or more stations running at various rates.
As one example, I did benchmark a ubnt UHD60ghz radio. It was definately FQ-ing, however it had such short queues as to cause slow start to exit very early and take a long time for tcp to ramp up to the actual bandwidth. 

Most of my battles to date have been elsewhere - starlink, 5G, cable, ethernet over powerline, etc. In the fiber market, really aggressive policers are present... I like cake's approaches, but am a bit biased.

**Comment by syadnom** [comment]
I use adtran and ufiber gpon products.  Their shapers suck.  They are
strict, but they suck.  Basically just getting stuck in the TDMA
distribution.  ufiber isn't sophisticated at all.  It's either a fifo
shaper on the ONT with an edgerouter at the edge, or it's round robin
on the tdma scheduler.  Adtran actually scheduled TDMA rx/tx windows
at the OLT for speed, but this is very much like a fifo buffer so cake
definitely does better.  I wouldn't call either one an AQM type
system.  I think the industries' opinion is that if you are delivering
gig speeds at low latency you don't need AQM.  I don't agree.

All the new 60Ghz gear is essentially shaper-free.  Ubiquiti's 'ad
products do a long buffer to do DSCP prioritization (poorly) and can
ramp up slowly in modulation because of some power saving schema
that's been re-worked/removed from beta firmwares.  The 'wave' gear
doesn't have this at all, instant response and it looks like fifo
buffers when you bury it.

In all of these scenarios, cake helps, but also many of us are
offering slower plans on this gear, 200x200 for example and it's much
nicer having that in cake than just hoping they never saturate the
service.

Since fq_codel and cake made it into routeros7, I've been flat abusing
it haha.  I have shapers on a number of my backup links so when
something goes wrong, cake catches  it and shapes at the backup
connection's capacity.  Just pointing this out because there are
scenarios other than end-user shaping in the mix.

I have noticed I get better results on routeros7 in a queue tree if
the parent shapers are hbt and the shaper targeting the customer is
cake.  ie, hbt-uplink/hbt-site/hbt-AP/cake-for-customer good,
cake/cake/cake/cake-for-customer not so good.

**Comment by dtaht** [comment]
htb-<htb<-htb<-cake is essentially a pull-through model, where at most one cburst worth of packets is transiting the htb layer
at any one time, and cake makes the decision as to what to queue up next. It is the best possible model at this time vs a vs cake, cake, cake. 

Also, in general (RFC7567) some form of fq+aqm belongs on every fast->slow transition, as you are doing today, and I wish more were. thing is, I'd hoped in mikrotik's case that they would be using BQL + fq_codel by default, as we do in linux and openwrt generally, instead of whatever they are doing. BQL queues up only one interrupt's worth of packets, and punts the rest of the work to fq_codel. This is 20x less cpu than shaping is, and (for example) at 100mbit, typically queues up only 3k.

**Comment by thebracket** [comment]
It needs a bunch of cleanup, but I've got `xdp-cpumap-tc` working with a `Trie` rather than a `hash map`. The source is in the master branch at: https://github.com/thebracket/xdp-cpumap-tc/

You can test by cloning the repo, and replacing the submodule with a link (`ln -s`) to my version. It's drop-in compatible, but `xdp_iphash_to_cpu_cmdline` now accepts an extra optional parameter `--prefix`. For example, `./xdp-cpumap-tc/src/xdp_iphash_to_cpu_cmdline --add --ip 192.168.100.0 --cpu 0 --classid 1:5 --prefix 24` will match everything in `192.168.100.0/24`.

I've only lightly tested it (with VMs talking to one another) - but it's working for me. Performance is still good, but I didn't heavily stress it.

I ran into one oddity: I had to reverse source/destination IPs in the `tc_classify` script. Before I did that, I was never getting any matches. Reversing it, and it matches consistently. (Don't reverse the config file!). Tested with libre 0.9 and 1.1 so far.

If anybody would like to give it a test (I don't recommend production testing, yet!), I'd be grateful.

> Good news: this puts me about 1/2 way to getting Ipv6 to work. I think I'll just extend the `Trie` key to a full 128-bits with a max search depth of 128, and put IPv4 in the 0000::ipv4 space - that used to be recommended for transition, and it avoids having to maintain separate maps (and bogging down the execution path with branches).

**Comment by interduo** [comment]
@thebracket maan You are awesome! Maybe You could do PR to root project?
I will test that but I need a little more time to make some lab for that purpose.

**Comment by thebracket** [comment]
I'd like to get a PR upstreamed, but I have to clean things up quite a bit before it's likely to be accepted. There's a bunch of notes lying around to help me discover what's going on!

**Comment by tohojo** [comment]
Feel free to re-use any of the code from the 'ipv6' branch of the original repo if that makes things easier! :)

**Comment by interduo** [comment]
> I'd like to get a PR upstreamed, but I have to clean things up quite a bit before it's likely to be accepted. There's a bunch of notes lying around to help me discover what's going on!

@thebracket I saw that You do big rewrite of xdp-project-tc: it better to sync in smaller parts because it would be easier to test and catch a bugs if any - maybe You could upload some code to mother project now

(maaan I saw the killer feature You did: support for trie design in XDP filtering with tc-cpumap-tc - that would be very big performance booster)

**Comment by thebracket** [comment]
@tohojo - I updated my fork to the latest, and after updating sub-modules and rebuilding I'm getting a *lot* of errors trying to use the thing?

I first ran into `libbpf.so.1` not found, so I ran `make install` and ended up symlinking it into `/usr/lib` (it went into lib64). That removed the first error, now I mostly see `libbpf: elf: legacy map definitions in 'maps' section are not supported by libbpf v1.0+` and it fails to build the maps completely. I'm in the process of applying the migrations from https://github.com/libbpf/libbpf/wiki/Libbpf:-the-road-to-v1.0 to my fork, but I think upstream may be borked.

**Comment by tohojo** [comment]
Ah, oops, didn't realise the map definitions were also the old format. Will fix those later tonight, and yeah, the error messages are... Not friendly :(

**Comment by dtaht** [comment]
For the record, tohojo is in CET, I am in PST, and I have no idea where everybody else is. His tonight is your tomorrow, or do I have that backwards?

**Comment by thebracket** [comment]
I'm in US central time, but in reality I'm on BabyTime (TM).... in other
words, I have no idea what time it is.

**Comment by tohojo** [comment]
Reverted the buggy conversion in upstream master, and will push a working version to a separate branch so it can be tested properly before merging this time :)

**Comment by dtaht** [comment]
tease.

**Comment by Belyivulk** [comment]
Ahahaha – I know this pain



HAYDEN SIMON
UBER GROUP LIMITED
MANAGING DIRECTOR - SUPREME OVERLORD
E: ***@***.***
M: 021 0707 014
W: www.uber.nz
53 PORT ROAD | PO BOX 5083 | WHANGAREI | NEW ZEALAND

**Comment by tohojo** [comment]
> tease.

Well, turned out that the conversion needed quite a bit more work than I had anticipated. PR now at https://github.com/xdp-project/xdp-cpumap-tc/pull/11

**Comment by thebracket** [comment]
The upstream patch is treating me very well. :-) I've been porting my Trie work to it as a test, and so far it's working better than before. (Skipping IPv4 altogether and using a unified IPv4/6 approach - so when it's done it should handle both)

**Comment by dtaht** [comment]
As a fork/exec takes 50us or so, and dumping a batch of netlink messages in much less, over here lochnair just pointed me at his inprogress in rust cake related netlink stuff: https://github.com/Lochnair/sqm-autorate-rust/blob/master/src/netlink.rs

**Comment by thebracket** [comment]
*goes to read about Netlink* - that looks quite interesting!

**Comment by dtaht** [comment]
Squirrel! The batch facility is also a speedup...

https://github.com/rchac/LibreQoS/issues/119

Anyway, if there is a sane way to create a netlink interface to htb and cake, there are a lot of constant definitions, TCA_whatever exported by the kernel, and I've always kind of wondered how to go about making that import sanely into rust.

 Also it pays to leverage the existing code (which is also GPL'ed). https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/tc/q_cake.c#n146

**Comment by interduo** [comment]
@thebracket Did You end up with trie support in xdp-cpumap-tc fork? Could push PR upstream?

**Comment by thebracket** [comment]
It's working with both ipv4 and ipv6. I'll make it available (and submit a
PR) later today.

**Comment by thebracket** [comment]
I'm happy to present: https://github.com/thebracket/xdp-cpumap-tc/tree/TrieV6

This branch (which will be a PR as soon as I've finished a test script):

* Replaced the `map_ip_hash` structure with a `BPF_MAP_TYPE_LPM_TRIE`. The key is a new `ip_hash_key` structure, which contains a prefix length and 128-bit address storage.
* The `xdp_iphash_to_cpu_cmdline` has gained an extra flag, `--prefix`. You can add an address with a /24 using `xdp_iphash_to_cpu_cmdline --add --ip 100.64.1.3 --cpu 0 --classid 1:5 --prefix 24`.
* If you don't include `--prefix`, it defaults to `/32` for IPv4 and `/128` for IPv6 --- preserving compatibility with the existing setup (no changes required if you don't want to use these features).
* The command-line also now understands IPv6 addresses. So `xdp_iphash_to_cpu_cmdline --add --ip fe80::215:5dff:fe64:4d05 --cpu 0 --classid 1:5 --prefix 64` will setup a matcher for that IPv6 address and all of its children.
* The command-line `--list` function has changed to also show the prefix, e.g. `"100.64.1.3/24" : { "cpu" : 0, "tc_maj" : "1" , "tc_min" : "5" },` or `"fe80::215:5dff:fe64:4d05/64" : { "cpu" : 0, "tc_maj" : "1" , "tc_min" : "5" }`
* The `clear` command no-longer takes the really odd route of translating an address into a string, and then translating it right back again for each address to remove it; it just removes all items.
* Once read, all addresses are turned into 128-bit addresses. IPv4 is placed in the last 4 octets of the address, with the rest zeroed. The command-line adds 96 to IPv4 prefixes automatically, so you don't need to remember to turn /24 into /120 (which it will be internally).
* The classifier uses a Trie lookup, starting with 128-bits (trying for an exact match). In testing, it seems very fast.

My test environment consists of 3 VMs: one running the shaper (I'm using LibreQOS to generate the basic commands), and a VM on either side of a bridge in the shaper. `iperf` shows that it is shaping correctly both for known and unknown addresses, and is correctly placing prefixed addresses in the right group.

Also, if you add this you may want to reboot to get rid of the previous pinned map. I didn't add any code to remove/replace it.

References: https://github.com/xdp-project/xdp-cpumap-tc/issues/4

I'll be sending this upstream as a PR shortly, I just want to finish a set of test scripts to make sure that the command-line tooling is bullet-proof.

**Comment by dtaht** [comment]
There is now a libreqos mailing list established over here: https://lists.bufferbloat.net/listinfo/libreqos

I am perhaps the last person in the world that prefers email, and it does seem that we can be highly productive on github, with pictures with circles and arrows and a paragraph on the back of each one, so if there is no or low interest (?), I will discontinue it.

**Comment by richb-hanover** [comment]
Another possibility might be the [Discussions](https://github.com/rchac/LibreQoS/discussions) facility that Github provides and that has already been enabled for LibreQoS...

**Comment by interduo** [comment]
> There is now a libreqos mailing list established over here: https://lists.bufferbloat.net/listinfo/libreqos
> 
> I am perhaps the last person in the world that prefers email, and it does seem that we can be highly productive on github, with pictures with circles and arrows and a paragraph on the back of each one, so if there is no or low interest (?), I will discontinue it.

We are in 2022 not 1999. Please ... :)

![image](https://user-images.githubusercontent.com/17087236/193028465-27620f68-089a-4acd-ab79-346e4260c3fe.png)

**Comment by thebracket** [comment]
I went ahead and joined the mailing list. Hopefully it'll be less likely to be considered spam (by Gmail) than the FreeBSD groups. Honestly, there's not a lot of difference between using discussions, a listserv, a never-ending issue on my end; every single one of them sends me email, and most of the time I reply to issues by email anyway.

**Comment by interduo** [comment]
@thebracket look at this carefully - if somebody is editing the posts You wont be notified

**Comment by tohojo** [comment]
Interduo ***@***.***> writes:

> @thebracket look at this carefully - if somebody is editing the posts
> You wont be notified

Which is why editing is a terrible idea, and anyone editing their post
should assume that some people won't read the edit (I also mostly
consume things via email) :)

**Comment by dtaht** [comment]
I like editing. I sometimes have grave difficulties with aphasia (which is why I don't use irc all that much), and it helps to hit save, and go back and edit mistakes and incorrect language, later. 

Btw, anyone for an irc channel?

/me hides

**Comment by thebracket** [comment]
> anyone for an irc channel?

I still find IRC to be a great way to find assistance, sometimes. Seems like half the world has switch to Discord, which is IRC with bad graphics IMO.

**Comment by interduo** [comment]
Modern developers just use telegram groups. This is much better than IRC ;)
btw Telegram got transport for IRC

**Comment by dtaht** [comment]
Matrix?

**Comment by rchac** [comment]
@thebracket 's IPv4/IPv6 trie pull request for xdp-cpumap-tc has been [approved](https://github.com/xdp-project/xdp-cpumap-tc/pull/14)! Can we all give a round of applause for @thebracket ?

**Comment by interduo** [comment]
@rchac is this a place for v1.3 version now?

**Comment by dtaht** [comment]
Heh. Was not the miracle of the juniper^Wipv6 berries enough? Can't we just go outside and enjoy a lovely friday afternoon and weekend first? Anyway... To me - classifying and closing the remaining bugs would be good, and starting a roadmap doc on the discussion page? Or here? Is there a way to give more folk the ability to label bugs?

**Comment by dtaht** [comment]
https://www.youtube.com/watch?v=UICh3ScfNWI went by today. It's great.... If only we could reach more ISPs. I've always been about the bottom up DIY fix the router thing, but libreqos is needed also to get it more right.

**Comment by syadnom** [comment]
Top level shapers w/ cake and a few more vocal wisps talking about it and I
think you're going to see a LOT of wisps come over.  the pping stuff and
getting those things into graphs will be a BIG factor on pulling in preseem
users.   The UISP integration is a big piece, but adding VISP is the next
step as the 2nd most popular CRM/NMS.  Also, maybe a graphical config and
topology interface, command line and text config scares a lot of people.

The real problem I think is that when these things come together you'll
have too much of a business on your hands... might need to make some sort
of 'certified libreqos support specialist' program lol

**Comment by dtaht** [comment]
I meant to talk to visp at the wispalaloosa show, got a contact there?

**Comment by syadnom** [comment]
***@***.*** or  ***@***.***

**Comment by dtaht** [comment]
hilariously, that's edited out. Could you drop me an email directly?

**Comment by dtaht** [comment]
"certified LibreQos support specialist" is possibly one way to go long term. I could see a package for doing install and network testing to review the current state of things, a free support tier, stuff like GTM 3.0. For some inspiration on that, see: https://tailscale.com/blog/free-plan/

**Comment by syadnom** [comment]
I just see the need for implementation and service support that is not really in the wheelhouse of most open source projects.  To get community support for general help is great, but there will be many that cannot grasp the 'linuxiness' of it and want a proper implementation.  That could be completely outside project... but a paid training and certificate would allow some monetization of that effective comercialization of the project.

**Comment by dtaht** [comment]
Another model: http://www.toad.com/gnu/cygnus/

**Comment by syadnom** [comment]
I'm game.  Primary issue for me is IPv6.  That's what keeps me on preseem.

Frankly, I could put money into the project for bounties if I could get
IPv6 support and replace preseem.

**Comment by thebracket** [comment]
Good news: my ipv6 support PR for xdp-cpumap is merged, so the backend
supports IPv6 now. (It supports IPv4 CIDR now, also}

**Comment by dtaht** [comment]
I would enjoy a chance for us all to get  together for a videoconference to discuss a roadmap. Thursdays are my big meeting day, but wednesdays are good too. To get NZ in the picture, somewhere between 12-2 PST (which is friday in NZ) is good. I generally use the galene videoconferencing server, and have them up all over the world. So I setup a new one here: 

https://spacewrt.net:8443/group/libreqos/ 

It works best with chrome, any login you want, no password. Would anyone here be able to get together in these windows this week?

**Comment by syadnom** [comment]
I'm game for whenever.  I don't know how much I'll contribute but I'll show
just in case.

**Comment by dtaht** [comment]
ok, how about 1PM PST thursday?

**Comment by thebracket** [comment]
Thursday, as in tomorrow? I think that will work.

**Comment by dtaht** [comment]
@rchac ? ?

**Comment by rchac** [comment]
Sorry for the delay, I have been overwhelmed with ISP work. Had to hire someone to help juggle all the paperwork for broadband grants. Things should be finally calming down now.

Sure thing, 1PM PST Thursday would work for me as well. =)

**Comment by dtaht** [comment]
I wanted to note that we did more or less establish a v1.3, v1.4, 2.0, 3.0 roadmap in that meeting but haven't really written up the results. Bracket (with toke and simon) is off making ebf pping fly, robert trying to make it more possible to do dynamic updates, and I'm off trying to get the rb5009 BQL fixed up to try an openwrt port on. We also agreed on a semi-sync-to-the linux kernel release schedule - every 6-8 weeks, probably offset by 2 weeks or so. We also agreed to meet via videoconference monthly. 

More discussions have been ongoing on the libreqos email list: https://lists.bufferbloat.net/listinfo/libreqos - please join. Please keep using github for bug reports and feature requests too! Despite those top two (and very hard) items being worked on, what else would y'all like to see (and/or contribute to) in 6 weeks?

**Comment by interduo** [comment]
> More discussions have been ongoing on the libreqos email list: https://lists.bufferbloat.net/listinfo/libreqos - please join. Please keep using github for bug reports and feature requests too! Despite those top two (and very hard) items being worked on, what else would y'all like to see (and/or contribute to) in 6 weeks?

All the mainly needed things are in issues. Good way is to put issues into Project called "Roadmap" (not existing now) on github or closed with #wontdo.

**Comment by thebracket** [comment]
I've been making some pretty solid headway on getting `pping` functionality into the classifiers, rather than relying on the `pping` utility (for anyone coming from Preseem, latency tracking is *really* important). There's a very much work-in-progress version available at https://github.com/thebracket/cpumap-pping-hackjob if anyone wants to play with it. (Before the inevitable "please merge to upstream" from @interduo , it's not ready for that - and upstream would prefer a more generic, less LibreQoS focused approach).

Testing between VMs (hyper-V on my desktop), I'm getting a pretty solid 4.5 gbit/s single-stream performance, while tracking TCP round-trip-times in a per-connection ringbuffer that spits out JSON in the format `{"tc":"1:5", "avg" : 5, "min" : 5, "max" : 5, "samples" : 1},` periodically).

I hope to test it on a live network (with a modified BracketQoS) at the weekend.

**Comment by syadnom** [comment]
Awesome guys.  I ended up in some wISP emergency work and couldn't make the meet.  Looking forward to the future.

**Comment by interduo** [comment]
@thebracket could You write a little bit more how Your code could be used in LibreQoS?

**Comment by rchac** [comment]
@thebracket 

I tested cpumap-pping in a VM test environment and it performed really well. With a modest amount of traffic with 4000 IPs in the trie - throughput was exactly the same and the CPU load vs xdp-cpumap-tc was almost exactly the same.

xdp-cpumap-tc:
![xdp-cpumap-tc](https://user-images.githubusercontent.com/22501920/197085033-673d9bd6-3a98-4512-9e1a-0dad4bbfd319.png)

cpumap-pping:
![cpumap-pping](https://user-images.githubusercontent.com/22501920/197085046-9ba21f82-e44d-4e15-a9e2-2961fd9970c4.png)

With cpumap-pping CPU use occasionally spikes very briefly, but it seems negligible. The latency results appear very consistent, and are super easy to parse from userspace. Really happy with this. Awesome work.

**Comment by thebracket** [comment]
Wow, you beat me to the punch! (I have Saturday early morning scheduled for
trying it with real-world data)

I'm very happy with your performance findings. I'll have to rename the
repo, it's not really a hackjob anymore (GitHub has a nice feature that if
you rename a repo, the old path still works - so I shouldn't break
anything). :-)

You have got me curious about the CPU spikes. I'll have to try and
instrument a local build and see if I can find the cause. One thing I did
find when testing the Trie stuff was that adding an explicit default (::/0)
and mapping it to a rule for unknown hosts seemed to help with consistency.
It avoids spitting out a log message, which is often a pretty slow process
(Linux's eBPF stuff is ludicrously fast, with timings in nanoseconds most
of the time).

**Comment by dtaht** [comment]
I am a big fan of comparison cdf plots and the *.flent.gz files being stashed somewhere. 

Also the rtt_fair_var tests can be used to talk to multiple hosts (virtual, or not). 

(thx for using flent!)

**Comment by dtaht** [comment]
Also I note this test *might* also show the minute differences in latency from different ethtool settings for the rx and tx rates if you squint at it, and run it long enough. y'all think 200usec for the EF flow is gooooood? hold my beer....

**Comment by dtaht** [comment]
Much of our traffic here has moved onto the mailing list of late. However, a place to discuss when the v1.3 release should go final, seems to be here. My guess is dec 7th would be a good target date? That gives everyone the holiday off, even if it slips a week. We also never got around to actually publishing the roadmap we'd discussed anywhere also!

**Comment by dtaht** [comment]
And while a ton of stuff has moved to the mailing list, "matrix" for r/t chat is being experimented with, please email robert for an invite. As I/we have achieved everything on this bug I laid out on sept 2....

"I have been watching this project expand in various directions rapidly and I hope, meeting many of your pent-up needs in production. I have some spare time in the coming quarter, and trying to get a grip on what is important, what is needed, what resources are missing, getting some bugs closed, and having an overall architectural discussion, and a sane roadmap is on my mind."

 THANK YOU EVERYONE FOR ALL YOUR ENTHUSIASM AND HARD WORK! TELL YOUR FRIENDS! (some help on designing a t-shirt would be nice) v.1.3 and beyond are going to be great! A business model would be nice to have too...

**Comment by interduo** [comment]
@rchac please invite me to matrix - my mail is `jkl@interduo.pl`
